pkg/ipcache/metadata_test.go

```
@@ -234,6 +235,35 @@ func TestOverrideIdentity(t *testing.T) {
 	assert.False(t, ok)
 }
 
+func TestUpsertMetadataTunnelPeerAndEncryptKey(t *testing.T) {
+	cancel := setupTest(t)
+	defer cancel()
+
+	ctx := context.Background()
+
+	IPIdentityCache.metadata.upsertLocked(worldPrefix, source.CustomResource, "node-uid",
+		types.TunnelPeer(netip.MustParseAddr("192.168.1.100")),
+		types.EncryptKey(7))
+	remaining, err := IPIdentityCache.InjectLabels(ctx, []netip.Prefix{worldPrefix})
+	assert.NoError(t, err)
+	assert.Len(t, remaining, 0)
+
+	ip, key := IPIdentityCache.getHostIPCache(worldPrefix.String())
+	assert.True(t, net.ParseIP("192.168.1.100").Equal(ip))
+	assert.Equal(t, uint8(7), key)
+
+	IPIdentityCache.metadata.upsertLocked(worldPrefix, source.CustomResource, "node-uid",
+		types.TunnelPeer(netip.MustParseAddr("192.168.1.200")))
+	IPIdentityCache.metadata.remove(worldPrefix, "node-uid", types.EncryptKey(7))
+	remaining, err = IPIdentityCache.InjectLabels(ctx, []netip.Prefix{worldPrefix})
+	assert.NoError(t, err)
+	assert.Len(t, remaining, 0)
+
+	ip, key = IPIdentityCache.getHostIPCache(worldPrefix.String())
+	assert.True(t, net.ParseIP("192.168.1.200").Equal(ip))
+	assert.Equal(t, uint8(0), key)
+}
```

I presume we should also check that `upsertLocked()` on the same IP with differing info should return an error?

---

(Basically just a request for additional test coverage based on the assumption that we can't get the same info from multiple sources --- though some of the discussion below seems to suggest that we may actually receive node updates for example from both k8s and etcd in some scenarios). Minor nit, not blocking the PR. We could always expand the unit test coverage after this PR.

---

Do you mean to assert that an upsert with a weaker source (that contains a different tunnel IP) does not change the original ipcache entry?

---

I tried messing around with this and I'm not getting the behavior I'm expecting. I applied this diff here (ignore the lines that switch from `worldPrefix` to `inClusterPrefix`):

<details><summary>Diff</summary>
<p>

```diff
diff --git a/pkg/ipcache/metadata_test.go b/pkg/ipcache/metadata_test.go
index b93cec72eb..e9db8b1d07 100644
--- a/pkg/ipcache/metadata_test.go
+++ b/pkg/ipcache/metadata_test.go
@@ -258,25 +258,37 @@ func TestUpsertMetadataTunnelPeerAndEncryptKey(t *testing.T) {
 
 	ctx := context.Background()
 
-	IPIdentityCache.metadata.upsertLocked(worldPrefix, source.CustomResource, "node-uid",
+	IPIdentityCache.metadata.upsertLocked(inClusterPrefix, source.CustomResource, "node-uid",
 		types.TunnelPeer(netip.MustParseAddr("192.168.1.100")),
 		types.EncryptKey(7))
-	remaining, err := IPIdentityCache.InjectLabels(ctx, []netip.Prefix{worldPrefix})
+	remaining, err := IPIdentityCache.InjectLabels(ctx, []netip.Prefix{inClusterPrefix})
 	assert.NoError(t, err)
 	assert.Len(t, remaining, 0)
 
-	ip, key := IPIdentityCache.getHostIPCache(worldPrefix.String())
+	ip, key := IPIdentityCache.getHostIPCache(inClusterPrefix.String())
 	assert.True(t, net.ParseIP("192.168.1.100").Equal(ip))
 	assert.Equal(t, uint8(7), key)
 
-	IPIdentityCache.metadata.upsertLocked(worldPrefix, source.CustomResource, "node-uid",
+	// Assert that an entry with a weaker source (and from a different
+	// resource) should fail or at least does not overwrite the existing
+	// (stronger) ipcache entry.
+	IPIdentityCache.metadata.upsertLocked(inClusterPrefix, source.Generated, "generated-uid",
+		types.TunnelPeer(netip.MustParseAddr("192.168.1.101")),
+		types.EncryptKey(7))
+	_, err = IPIdentityCache.InjectLabels(ctx, []netip.Prefix{inClusterPrefix})
+	assert.Error(t, err)
+	ip, key = IPIdentityCache.getHostIPCache(inClusterPrefix.String())
+	assert.True(t, net.ParseIP("192.168.1.100").Equal(ip))
+	assert.Equal(t, uint8(7), key)
+
+	IPIdentityCache.metadata.upsertLocked(inClusterPrefix, source.CustomResource, "node-uid",
 		types.TunnelPeer(netip.MustParseAddr("192.168.1.200")))
-	IPIdentityCache.metadata.remove(worldPrefix, "node-uid", types.EncryptKey(7))
-	remaining, err = IPIdentityCache.InjectLabels(ctx, []netip.Prefix{worldPrefix})
+	IPIdentityCache.metadata.remove(inClusterPrefix, "node-uid", types.EncryptKey(7))
+	remaining, err = IPIdentityCache.InjectLabels(ctx, []netip.Prefix{inClusterPrefix})
 	assert.NoError(t, err)
 	assert.Len(t, remaining, 0)
 
-	ip, key = IPIdentityCache.getHostIPCache(worldPrefix.String())
+	ip, key = IPIdentityCache.getHostIPCache(inClusterPrefix.String())
 	assert.True(t, net.ParseIP("192.168.1.200").Equal(ip))
 	assert.Equal(t, uint8(0), key)
 }
```

</p>
</details> 

The behavior I'm seeing is that an entry with the weaker source is actually overwriting the existing entry. In this case, a different tunnel IP overwrites the existing one, which...doesn't seem good. I'm not sure if we've fundamentally missed something here in the merging logic. :worried:

---

I believe that's caused by the code in my comment below:
https://github.com/cilium/cilium/pull/23208#discussion_r1139431586

---

Ah, you're right. I also needed to ensure that the [above code](https://github.com/cilium/cilium/pull/23208#discussion_r1139431586) ran against the prefix infos that are sorted by source. If we don't sort by source, we get the wrong behavior as mentioned in my previous msg.